src/mame/jaleco/jaleco_vj_qtaro.cpp

```diff
@@ -0,0 +1,653 @@
+{
+    auto r = m_int & ~0x10;
+
+    if (m_dma_running[0] || m_dma_running[1] || m_dma_running[2]) {
+        // The only time 0x10 is referenced is when ending WriteStream for the individual Qtaro devices.
+        // All 3 of the WriteStream cleanup functions start by writing 0 to qtaro_dma_requested_w and qtaro_dma_running_w
+        // then loop until 0x10 is not set here.
+        r |= 0x10;
+    }
+
+    return r;
+}
+
+void jaleco_vj_king_qtaro_device::int_w(offs_t offset, uint32_t data)
+{
+    // 0x1000000 is used to trigger an event interrupt in the Qtaro driver.
+    // The interrupt will only be accepted and cleared when event_r, event2_r, event_io_r are non-zero.
+    // It's set, read, and cleared all in the device driver on the PC so no need to handle it here.
+    m_int = data;
+}
+
+uint32_t jaleco_vj_king_qtaro_device::int_fpga_r(offs_t offset)
+{
+    return m_int_fpga;
+}
+
+void jaleco_vj_king_qtaro_device::int_fpga_w(offs_t offset, uint32_t data)
+{
+    m_int_fpga = data;
+}
+
+template <int device_id>
+void jaleco_vj_king_qtaro_device::qtaro_dma_requested_w(offs_t offset, uint32_t data)
+{
+    m_dma_running[device_id] = data == 1;
+
+    if (data == 1 && m_dma_descriptor_requested_addr[device_id] != 0) {
+        m_dma_descriptor_addr[device_id] = m_dma_descriptor_requested_addr[device_id];
+        m_dma_descriptor_length[device_id] = 0;
+        m_dma_descriptor_requested_addr[device_id] = 0;
+    }
+
+    LOGMASKED(LOG_DMA, "%lf %s qtaro_dma_requested_w<%d>: %08x\n", machine().time().as_double(), machine().describe_context().c_str(), device_id, data);
+}
+
+template <int device_id>
+void jaleco_vj_king_qtaro_device::qtaro_dma_descriptor_phys_addr_w(offs_t offset, uint32_t data)
+{
+    LOGMASKED(LOG_DMA, "qtaro_dma_descriptor_phys_addr_w<%d>: %08x %d\n", device_id, data, m_dma_running[device_id]);
+    m_dma_descriptor_requested_addr[device_id] = data;
+}
+
+template <int device_id>
+uint32_t jaleco_vj_king_qtaro_device::qtaro_dma_running_r(offs_t offset)
+{
+    return m_dma_running[device_id];
+}
+
+template <int device_id>
+void jaleco_vj_king_qtaro_device::qtaro_dma_running_w(offs_t offset, uint32_t data)
+{
+    LOGMASKED(LOG_DMA, "qtaro_dma_running_w<%d>: %08x\n", device_id, data);
+
+    if (data == 0)
+        m_dma_running[device_id] = false;
+}
+
+TIMER_CALLBACK_MEMBER (jaleco_vj_king_qtaro_device::video_dma_callback)
+{
+    // TODO: VJ does a weird thing where it'll stop the other two video streams when it
+    // sees that one video stream has finished reading its file, even if the other two video
+    // streams are still reading their files. How to fix that without breaking Stepping Stage?
+    // Setting the buffer size <= 0x8000 fixes it to some degree because the DMAs can't run
+    // ahead of each other too far, but there's an issue with Stepping Stage where it'll
+    // just not send data for a small period causing videos to pause and then go out of sync (???).
```

Comment:
First line of a `TODO` is supposed to be brief, this will cut off at first line at "when it" in an IDE viewer widget such as  [Todo Tree](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree)